### PR TITLE
fix: EphemeralFiles can outlive their Timeline via `enum LayerManager`

### PR DESCRIPTION
**Pull request by koivunej**
Ephemeral files cleanup on drop but did not delay shutdown, leading to problems with restarting the tenant. The solution is as proposed:
- make ephemeral files carry the gate guard to delay `Timeline::gate` closing
- flush in-memory layers and strong references to those on `Timeline::shutdown`

The above are realized by making LayerManager an `enum` with `Open` and `Closed` variants, and fail requests to modify `LayerMap`.

Additionally:

- fix too eager anyhow conversions in compaction
- unify how we freeze layers and handle errors
- optimize likely_resident_layers to read LayerFileManager hashmap values instead of bouncing through LayerMap

Fixes: #7830

### CODE REVIEW

File: pageserver/src/tenant/timeline/layer_manager.rs [hunk 1]
```diff
@@ -69,21 +69,14 @@ impl LayerManager {
...
     ) -> Result<Arc<InMemoryLayer>> {
         ensure!(lsn.is_aligned());
 
-        ensure!(
-            lsn > last_record_lsn,
-            "cannot modify relation after advancing last_record_lsn (incoming_lsn={}, last_record_lsn={})",
-            lsn,
-            last_record_lsn,
-        );
-
```
Comment (koivunej) [inline pageserver/src/tenant/timeline/layer_manager.rs:86]: I just moved this back to timeline (only caller) because that's where the last_record_lsn is to avoid a `#[allow(clippy::too_many_args)]`.

**Comment by koivunej** [comment]
I am overall worried how difficult are these enterings to debug. I was thinking of the list/collection/set in Gate but that would not really have told me much (inmemlayer was holding it). I would have needed a tool to show that this arc pointer is reachable through timeline -> timeline::write_lock -> TimelineWriterState::open_layer.

File: pageserver/src/tenant/timeline/layer_manager.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/tenant/timeline/layer_manager.rs:86]: I think the ensure makes more sense here than in the caller. Just slap on the allow(too_many_arguments), why not?

File: pageserver/src/tenant/timeline/layer_manager.rs [hunk 2]
```diff
@@ -208,6 +208,19 @@ impl LayerManager {
...
+    ///
+    /// If there was a want to flush in-memory layers, it must have happened earlier.
+    pub(crate) fn drop_inmemory_layers(&mut self, writer_state: &mut Option<TimelineWriterState>) {
+        let open = self.layer_map.open_layer.take();
+        let frozen = self.layer_map.frozen_layers.len();
+        self.layer_map.frozen_layers.clear();
+        assert_eq!(open.is_some(), writer_state.is_some());
+        writer_state.take();
+        tracing::info!(open = open.is_some(), frozen, "dropped inmemory layers");
+    }
```
Comment (problame) [inline pageserver/src/tenant/timeline/layer_manager.rs:222]: By simply taking out the in-memory layers, Timeline::get will be giving out wrong answers because we're not resetting last_record_lsn.

The only thing that prevents us from actually giving out the wrong answer is that we check for the cancellation token.
Too brittle for my taste.

So, I would much rather introduce a well-defined "we're shut down" state for the `LayerManager`, lik `enum LayerManager { Open(...), ShutDown }`.

**Review by problame — COMMENTED** [review]
IMO better PR title would be

```
fix: EphemeralFile objects outlive their Tenant
```

---

please undo the movement of the `ensure!`

---

File: pageserver/src/tenant/timeline/layer_manager.rs (hunk 1 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline/layer_manager.rs:86]: It is irresponsible to spray allow lints everywhere without consideration. Timeline has the last_record_lsn, so I think it should enforce this.. But you don't think that is a good enough barrier because Timeline::layers is pub(crate) or pub? I kind of agree.

File: pageserver/src/tenant/timeline/layer_manager.rs (hunk 2 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline/layer_manager.rs:222]: Timeline cancellation has been fired already and we take the write lock..

It think that should be enough.

Comment (problame) [reply pageserver/src/tenant/timeline/layer_manager.rs:222]: If we held the write lock on layer map until after the gate is closed an argument could be made that it's ok.

But, the reality is that we can always have places where we forget to check cancellation / gate, and if we access the layer map in those places, we return incorrect data.

I would sleep much better knowing there is robustness in depth.

Your call, I made my point.

Comment (koivunej) [reply pageserver/src/tenant/timeline/layer_manager.rs:222]: Yeah I do get your point. I'll try to convince myself more strongly. My expectation was that we would always have:

```
loop {
  // check cancellation
  // read_lock(search for remaining keys)
  // get_value_reconstruct_data(remaining keys, found layers)
  // maybe_continue
}
```

> defence in depth

I'd also like to prevent something else from getting worse: how the cancellations are spread around so wide right now. I am convinced we should not be checking for cancellations except for a very few places.

But as everything is so spread around maybe defense in depth is the way to go.

Comment (koivunej) [reply pageserver/src/tenant/timeline/layer_manager.rs:222]: There is at least the timeline_detail doing layer map accesses across the test suite after/at shutdown.

Comment (koivunej) [reply pageserver/src/tenant/timeline/layer_manager.rs]: Right now the many functions are not yet returning the Shutdown error, which would for example take precedence over RemoteTimelineClient's stopped error during late stages of compaction. Instead, there's `warn!` logging to give me (a lot of) test reports to verify.

Returning the `Result<(), Shutdown>` is no problem, and it is already handled even better than the current handling of the anyhow from RemoteTimelineClient.

Comment (koivunej) [reply pageserver/src/tenant/timeline/layer_manager.rs]: And I forgot to check before pushing... Well... Let's see the version with `Result<(), Shutdown>` then.

File: pageserver/src/tenant/timeline.rs [hunk 3]
```diff
@@ -3735,6 +3784,7 @@ impl Timeline {
             guard
                 .try_freeze_in_memory_layer(at, &self.last_freeze_at, write_lock)
                 .await
+                .expect("layer manager cannot be shutdown, because we are holding the write_lock")
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs]: I failed to clean this up just by itself, instead of trimming down on "fix: unify freeze+signal" which unifies the jungle of different methods for freezing, signalling flush after freeze, signalling flush after non-freeze, and maybe waiting for the flush loop to reach requested count.

**Comment by koivunej** [comment]
Created a take2 version of this PR, rationale: https://github.com/neondatabase/neon/pull/8332#issuecomment-2219882645

This PR still has good fixes like the anyhows, optimization for `likely_resident_layers` etc. but I will try to eventually commit them separately.

File: pageserver/src/tenant/timeline/layer_manager.rs [hunk 4]
```diff
@@ -24,66 +24,268 @@ use crate::{
...
 }
 
 impl LayerManager {
     pub(crate) fn get_from_desc(&self, desc: &PersistentLayerDesc) -> Layer {
-        self.layer_fmgr.get_from_desc(desc)
+        use LayerManager::*;
+        let maybe = match self {
+            Open(open) => open.get_from_desc(desc),
+            Closed { layers } => layers.get(&desc.key()),
```
Comment (problame) [inline pageserver/src/tenant/timeline/layer_manager.rs:49]: We should return an `Err(ShuttingDown)` here, no?

File: pageserver/src/tenant/timeline/layer_manager.rs [hunk 5]
```diff
@@ -24,66 +24,268 @@ use crate::{
...
+
+    pub(crate) fn all_persistent_layers(&self) -> Vec<PersistentLayerKey> {
+        self.layers().keys().cloned().collect_vec()
+    }
+
+    fn layers(&self) -> &HashMap<PersistentLayerKey, Layer> {
+        use LayerManager::*;
+        match self {
+            Open(OpenLayerManager { layer_fmgr, .. }) => &layer_fmgr.0,
+            Closed { layers } => layers,
```
Comment (problame) [inline pageserver/src/tenant/timeline/layer_manager.rs:245]: This here as well. We shouold make this return `Err(ShuttingDown)`

**Review by problame — COMMENTED** [review]
Thanks for putting in the effort to switch to an enum.

Let's have a call to discuss why in some places we continue granting access to the `layers` after shutdown.

**Comment by problame** [comment]
We had the call, I now understand the rationale for not failing `LayerManager::layers()` and callers after we've shut down the `LayerManager`.

I really like this impl, will do more review on it but I think this is the way to go because of the robustness-in-depth it provides.

In contrast, the alternative idea in https://github.com/neondatabase/neon/pull/8332 is to
use a separate gate for `EphemeralFile` so we can order the removal of the `InMemoryLayer` from layer manager strictly _after_ the Timeline::gate has closed.
That's quite clever, but doesn't have the same robustness-in-depth.

To the annoyance of everyone involved, I would like to entertain a third option: instead of making the LayerManager have a shut-down state, we could also push that down into the InMemoryLayer: make _it_ an `enum { Open(...), Frozen(...), ShuttingDown }`.
I can imagine it would result in slightly fewer error bubble ups required, because we only interact with InMemoryLayer on the data path, and data path already has `PageReconstructError::Cancelled`.
Sorry I didn't think about this earlier.

**Comment by koivunej** [comment]
> To the annoyance of everyone involved, I would like to entertain a third option: instead of making the LayerManager have a shut-down state, we could also push that down into the InMemoryLayer: make _it_ an `enum { Open(...), Frozen(...), ShuttingDown }`.
> I can imagine it would result in slightly fewer error bubble ups required, because we only interact with InMemoryLayer on the data path, and data path already has `PageReconstructError::Cancelled`.

The difference between the implemented solution alternatives and the above is that this version does not allow read operations to complete (detect cancellation or succeed with the read). The alternative with the additional gate lets them complete.

Humoring the proposed third option:
- currently the frozen-ness is handled via an `OnceCell` (not enum)
- in this solution all of the in-mem layers would be cleaned up with this state transition?
    - would require moving the cleanup from drop to state transition, no problem there

----

I'll cleanup the conflict soon-ish and ask a wider opinion.

File: pageserver/src/tenant/timeline.rs [hunk 6]
```diff
@@ -3133,14 +3173,13 @@ impl Timeline {
...
-            let historic_layer_name = historic_layer.layer_name();
-            if layer_name == &historic_layer_name {
-                return Some(guard.get_from_desc(&historic_layer));
-            }
-        }
-
-        None
+        let layer = guard
+            .layer_map()
+            .ok()?
```
Comment (jcsp) [inline pageserver/src/tenant/timeline.rs:3178]: The original concern is that we must avoid incorrectly concluding a layer doesn't exist during shutdown.  So although this function is not actually used in page reconstruction afaik, isn't it generally unsafe to conflate "layer not found" and "layer map not available"?

File: pageserver/src/tenant/timeline/layer_manager.rs [hunk 7]
```diff
@@ -24,66 +24,268 @@ use crate::{
...
+    }
+
+    /// Called when a GC-compaction is completed.
+    pub(crate) fn finish_gc_compaction(
+        &mut self,
+        compact_from: &[Layer],
+        compact_to: &[ResidentLayer],
+        metrics: &TimelineMetrics,
+    ) -> Result<(), Shutdown> {
+        self.open_mut()?
```
Comment (jcsp) [inline pageserver/src/tenant/timeline/layer_manager.rs:184]: These wrappers feel unnecessary: callers already have to be aware of the concept that their LayerManager might not be initialized (i.e. handling `Shutdown`), so it isn't onerous to also make the caller do a `layers.open_mut()?.finish_gc_compaction` rather than maintaining this set of 2 line wrappers

**Comment by jcsp** [comment]
I played with moving the `enum{}` wrapper up into Timeline, and it works, but the bulk of the code churn remains of course in changing all the infallible code paths to be fallible, so it's not much of a net benefit.

In future, what we really need is some ActiveTimeline guard that holds a gate and ensures the interior timeline is not shut down, so that it can do infallible access to the layer map.  An outer class that implements the "stuff that only makes sense if a timeline is active" stuff could be a step in the direction of breaking up Timeline a bit (in future))

File: pageserver/src/tenant/timeline.rs (hunk 6 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3178]: I think this API exists only to fulfill manual layer eviction. But you are right, it might screw up a benchmark, I will re-check this.

Comment (koivunej) [reply pageserver/src/tenant/timeline.rs]: [521fbd2](https://github.com/neondatabase/neon/pull/8229/commits/521fbd2674f29598d2b15080c532560156eae33c)

File: pageserver/src/tenant/timeline/layer_manager.rs (hunk 7 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline/layer_manager.rs:184]: Yes, pushing this soon.

Comment (koivunej) [reply pageserver/src/tenant/timeline/layer_manager.rs]: 077a9a4a4

File: pageserver/src/tenant/timeline.rs [hunk 8]
```diff
@@ -1898,6 +1924,11 @@ impl Timeline {
                     // about corner cases like s3 suddenly hanging up?
                     self.remote_client.shutdown().await;
                 }
+                Err(FlushLayerError::Cancelled) => {
+                    // this is likely the second shutdown, ignore silently.
+                    // TODO: this can be removed once https://github.com/neondatabase/neon/issues/5080
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:1929]: 5080 is marked as closed. Should it be reopened?

File: pageserver/src/tenant/timeline.rs [hunk 9]
```diff
@@ -1923,10 +1955,16 @@ impl Timeline {
...
-        // TODO: work toward making this a no-op. See this funciton's doc comment for more context.
+        // TODO: work toward making this a no-op. See this function's doc comment for more context.
         tracing::debug!("Waiting for tasks...");
         task_mgr::shutdown_tasks(None, Some(self.tenant_shard_id), Some(self.timeline_id)).await;
 
+        {
+            // Allow any remaining in-memory layers to do cleanup.
+            let mut write_guard = self.write_lock.lock().await;
+            self.layers.write().await.shutdown(&mut write_guard);
+        }
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:1966]:
```suggestion
        {
            // `InMemoryLayer`s keep the timeline Gate open.
            // => destroy them before waiting for gate closure below.
            let mut write_guard = self.write_lock.lock().await;
            self.layers.write().await.shutdown(&mut write_guard);
        }
```

File: pageserver/src/tenant/timeline.rs [hunk 10]
```diff
@@ -3724,21 +3784,48 @@ impl Timeline {
...
+        let mut my_flush_request = 0;
+
+        let flush_loop_state = { *self.flush_loop_state.lock().unwrap() };
+        if !matches!(flush_loop_state, FlushLoopState::Running { .. }) {
+            return Err(FlushLayerError::NotRunning(flush_loop_state));
+        }
+
+        self.layer_flush_start_tx.send_modify(|(counter, lsn)| {
+            my_flush_request = *counter + 1;
+            *counter = my_flush_request;
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:3822]: checked_add and panic if check fails? Otherwise the assert below "will" fail and it will be unclear why

File: pageserver/src/tenant/timeline.rs [hunk 11]
```diff
@@ -3724,21 +3784,48 @@ impl Timeline {
...
+            *lsn = std::cmp::max(at, *lsn);
+        });
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:3824]: No objections to this code but, was it _necessary_ for this PR or a drive-by improvement?

Comment (problame) [reply pageserver/src/tenant/timeline.rs:3824]: Ah nvmd this is just inlining existing code from removed function `flush_frozen_layers`

File: pageserver/src/tenant/timeline.rs [hunk 12]
```diff
@@ -4062,17 +4121,11 @@ impl Timeline {
...
-            assert!(new_value >= old_value);
-            self.disk_consistent_lsn.store(new_value);
-            true
-        } else {
-            false
-        }
+        let old_value = self.disk_consistent_lsn.fetch_max(new_value);
+        assert!(new_value >= old_value);
+        new_value != old_value
     }
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:4129]: This code is asserting _after_ the damage is done. Why not keep the current code?

File: pageserver/src/tenant/timeline.rs [hunk 13]
```diff
@@ -4663,8 +4718,11 @@ impl Timeline {
         let mut guard = self.layers.write().await;
 
         // FIXME: we could add the images to be uploaded *before* returning from here, but right
-        // now they are being scheduled outside of write lock
-        guard.track_new_image_layers(&image_layers, &self.metrics);
+        // now they are being scheduled outside of write lock; current way is inconsistent with
+        // compaction lock order.
+        guard
+            .open_mut()?
+            .track_new_image_layers(&image_layers, &self.metrics);
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:4725]: How confident are you this won't cause deadlocks?

File: pageserver/src/tenant/timeline.rs [hunk 14]
```diff
@@ -4866,7 +4932,7 @@ impl Timeline {
                 // because we have not implemented L0 => L0 compaction.
                 duplicated_layers.insert(l.layer_desc().key());
             } else if LayerMap::is_l0(&l.layer_desc().key_range) {
-                bail!("compaction generates a L0 layer file as output, which will cause infinite compaction.");
+                return Err(CompactionError::GeneratedL0);
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:4935]: This can be manual `CompactionError::Other()`. Nobody cares about the difference between `CompactionError::GeneratedL0` and `CompactionError::Other` => just needless `match` bureaucracy at the places that inspect `CompationError`s.

File: pageserver/src/tenant/timeline.rs [hunk 15]
```diff
@@ -5918,9 +5997,10 @@ impl<'a> TimelineWriter<'a> {
         // self.write_guard will be taken by the freezing
         self.tl
             .freeze_inmem_layer_at(freeze_at, &mut self.write_guard)
-            .await;
+            .await
+            .context("roll layer after batch")?;
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:6001]: We can avoid the anyhow here

**Comment by problame** [comment]
> In future, what we really need is some ActiveTimeline guard that holds a gate and ensures the interior timeline is not shut down, so that it can do infallible access to the layer map. An outer class that implements the "stuff that only makes sense if a timeline is active" stuff could be a step in the direction of breaking up Timeline a bit (in future))

CF the `Handle` added in https://github.com/neondatabase/neon/pull/8339/files#diff-903846e8f4524450320e7bd111f5fb36159867c47edab206b5b3ed8e62f85f48R81-R93

File: pageserver/src/tenant/timeline.rs (hunk 11 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3824]: Yeah. It was at least quite difficult for me to see how to do it without fixing the remaining parts.

File: pageserver/src/tenant/timeline.rs (hunk 12 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:4129]: fetch_max does not modify the value if the stored value is larger than the given value, so no damage did get to happen.

I did split this out to a separate PR. i thought it is a footgun to have this non-atomic lockless code lying around which has 2 callsites.

File: pageserver/src/tenant/timeline.rs (hunk 13 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:4725]: it has been like this for a long time, I am doing somw drive by comment adjustment here.

File: pageserver/src/tenant/timeline.rs (hunk 8 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:1929]: It is one of these class issues... I made a note there. I think it doesn't really matter what are we targeting, as long as we are consistent what issue are we targeting.

File: pageserver/src/tenant/timeline.rs (hunk 12 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:4129]: I've already split this to another PR #8311.

File: pageserver/src/tenant/timeline.rs (hunk 9 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:1966]: Unsure why this is needed. There will be a lockup if this is skipped.

File: pageserver/src/tenant/timeline.rs (hunk 14 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:4935]: I see you already changed it in another PR :D I removed it.

File: pageserver/src/tenant/timeline.rs (hunk 15 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs]: Avoided in [3b3b54f](https://github.com/neondatabase/neon/pull/8229/commits/3b3b54fcc106e2ba50cbe1a64e78e09ce26e1472)

File: pageserver/src/tenant/timeline.rs (hunk 9 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs]: [ff459df](https://github.com/neondatabase/neon/pull/8229/commits/ff459dfd4ac773a68ead92c7baa3e537bc0bf2e3)

File: pageserver/src/tenant/timeline.rs (hunk 11 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3379]: https://github.com/neondatabase/neon/pull/8229/commits/288d9eb9b4053d58dd6f2abbc61fda0fb547d796

File: pageserver/src/tenant/timeline.rs [hunk 16]
```diff
@@ -2953,6 +3001,7 @@ impl Timeline {
 }
 
 impl Timeline {
+    #[allow(unknown_lints)] // doc_lazy_continuation is still a new lint
```
Comment (koivunej) [inline pageserver/src/tenant/timeline.rs:3004]: rebase failure
```suggestion
```

File: pageserver/src/tenant/layer_map.rs [hunk 17]
```diff
@@ -846,8 +846,8 @@ impl LayerMap {
     }
 
     /// Return all L0 delta layers
-    pub fn get_level0_deltas(&self) -> Vec<Arc<PersistentLayerDesc>> {
-        self.l0_delta_layers.to_vec()
+    pub fn level0_deltas(&self) -> &Vec<Arc<PersistentLayerDesc>> {
```
Comment (koivunej) [inline pageserver/src/tenant/layer_map.rs:849]: This should probably be a slice.

**Review by arpad-m — APPROVED** [review]
we went over this PR in a 1:1 call

File: pageserver/src/tenant/timeline.rs [hunk 18]
```diff
@@ -3534,11 +3518,6 @@ impl Timeline {
         }
     }
 
-    async fn flush_frozen_layers_and_wait(&self, at_lsn: Lsn) -> Result<(), FlushLayerError> {
-        let token = self.flush_frozen_layers(at_lsn)?;
-        self.wait_flush_completion(token).await
-    }
-
```
Comment (koivunej) [inline pageserver/src/tenant/timeline.rs:3460]: So this was inlined because it had three jobs, 2 callsites:
- freeze
- signal that we've flushed
- wait for flush

so now there's less methods:
- one which freezes and signals
- one which waits

You are correct, it was not really necessary to remove this, however, `flush_frozen_layers_and_wait` didn't freeze, it only signalled, and I thought it was stupid.

File: pageserver/src/tenant/timeline.rs [hunk 19]
```diff
@@ -1502,29 +1508,9 @@ impl Timeline {
...
-                    // it is possible the LayerManager shutdown happened between us releasing the
-                    // read lock and re-acquiring the write lock; in such case, silently do
-                    // nothing.
-
-                    let res = match res {
-                        Ok(true) => self.flush_frozen_layers(current_lsn).map(|_req| ()),
-                        Ok(false) => Ok(()),
-                        Err(e) => Err(e.into()),
-                    };
```
Comment (koivunej) [inline pageserver/src/tenant/timeline.rs]: So all of this was: `freeze and if frozen, then signal`, which is now in `freeze_inmem_layer_at`.

File: pageserver/src/tenant/timeline.rs [hunk 20]
```diff
@@ -1451,9 +1450,16 @@ impl Timeline {
                     );
 
                     // The flush loop will update remote consistent LSN as well as disk consistent LSN.
-                    self.flush_frozen_layers_and_wait(last_record_lsn)
-                        .await
-                        .ok();
```
Comment (koivunej) [inline pageserver/src/tenant/timeline.rs:1429]: For this it doesn't matter that there is nothing to freeze, we will still signal and then we opt-in to waiting.

File: pageserver/src/tenant/timeline.rs [hunk 21]
```diff
@@ -3360,23 +3346,48 @@ impl Timeline {
...
-                .expect("layer manager cannot be shutdown, because we are holding the write_lock")
+                .open_mut()?
                 .try_freeze_in_memory_layer(at, &self.last_freeze_at, write_lock)
                 .await
         };
+
         if frozen {
             let now = Instant::now();
             *(self.last_freeze_ts.write().unwrap()) = now;
         }
```
Comment (koivunej) [inline pageserver/src/tenant/timeline.rs:3370]: I think before with the two different ways, we could had had a freeze without having a `self.last_freeze_ts` updated. However, it might be I misremember, and already fixed that.

However, I think this is now much better, this was really a big mess.

File: pageserver/src/tenant/timeline.rs [hunk 22]
```diff
@@ -3360,23 +3346,48 @@ impl Timeline {
...
+
+        let flush_loop_state = { *self.flush_loop_state.lock().unwrap() };
+        if !matches!(flush_loop_state, FlushLoopState::Running { .. }) {
+            return Err(FlushLayerError::NotRunning(flush_loop_state));
+        }
+
+        self.layer_flush_start_tx.send_modify(|(counter, lsn)| {
+            my_flush_request = *counter + 1;
+            *counter = my_flush_request;
+            *lsn = std::cmp::max(at, *lsn);
```
Comment (arpad-m) [inline pageserver/src/tenant/timeline.rs:3386]:
```rust
            *lsn = at.max(*lsn);
```

**Comment by koivunej** [comment]
We also had a second session on the unified signal refactoring.
